lib/apiservers/engine/backends/filter/container.go

```diff
@@ -0,0 +1,230 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filter
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/docker/engine-api/types"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/cache"
+	"github.com/vmware/vic/lib/apiservers/portlayer/models"
+)
+
+// reused from docker/docker/daemon/list.go
+type ContainerListContext struct {
+	FilterContext
+
+	// idx is the container iteration index for this context
+	Counter int
+	// ExitCode for the passed container
+	ExitCode int
+	// exitAllowed is a list of exit codes allowed to filter with
+	exitAllowed []int
+	// ContainerListOptions is the filters set by the user
+	*types.ContainerListOptions
+}
+
+/*
+*	IncludeContainer will evaluate the filter criteria in listContext against the provided
+* 	container and determine what action to take.  There are three options:
+*		* IncludeAction
+*		* ExcludeAction
+*		* StopAction
+*
+ */
+func IncludeContainer(listContext *ContainerListContext, container *models.ContainerInfo) FilterAction {
+
+	// filter common requirements
+	act := filterCommon(&listContext.FilterContext)
+	if act != IncludeAction {
+		return act
+	}
+
+	// Do not include container if the name doesn't match
+	for i := range container.ContainerConfig.Names {
+		if !listContext.filters.Match("name", container.ContainerConfig.Names[i]) {
+			return ExcludeAction
+		}
+	}
+
+	// Stop iteration when the index is over the limit
+	if listContext.Limit > 0 && listContext.Counter == listContext.Limit {
+		return StopAction
+	}
+
+	// Do not include container if its exit code is not in the filter
+	if len(listContext.exitAllowed) > 0 {
+		shouldSkip := true
+		for _, code := range listContext.exitAllowed {
+			// note "Running" state is congruent with PortLayer and not docker
+			if code == listContext.ExitCode && container.ContainerConfig.State != "Running" && container.ProcessConfig.StartTime > 0 {
+				shouldSkip = false
+				break
+			}
+		}
+		if shouldSkip {
+			return ExcludeAction
+		}
+	}
+
+	state := DockerState(container.ContainerConfig.State)
+	// Do not include container if its status doesn't match the filter
+	if !listContext.filters.Match("status", state) {
+		return ExcludeAction
+	}
+
+	// filter on networking
+	networkExist := fmt.Errorf("container part of network")
+	for n := range container.Endpoints {
+		name := container.Endpoints[n].Scope
+		// TODO: this is the incorrect ID
+		id := container.Endpoints[n].ID
+		if listContext.filters.Include("network") {
+			err := listContext.filters.WalkValues("network", func(value string) error {
+				if name == value || id == value {
+					return networkExist
+				}
+				return nil
+			})
+			if err != networkExist {
```

Comment:
Dug into this more -- the code is misleading -- it's not really concerned about error equality it's concerned about nil vs. !nil -- I'm updating the code to make that obvious.  Thanks for pointing it out.